Code size when using `DEFMT_LOG`

I was trying to figure out why my binary had gotten so huge when I realized that `DEFMT_LOG` is causing a massive blowup in code size. With `DEFMT_LOG=info,discobiker::actors::display=trace`, `cargo expand` shows that every single defmt macro invocation in my package is compiling this `check` function:

```rust
const fn check() -> bool {
    let module_path = "discobiker".as_bytes();
    if if 27usize > module_path.len() {
        false
    } else {
        module_path[0usize] == 100u8
            && module_path[1usize] == 105u8
            && module_path[2usize] == 115u8
            && module_path[3usize] == 99u8
            && module_path[4usize] == 111u8
            && module_path[5usize] == 98u8
            && module_path[6usize] == 105u8
            && module_path[7usize] == 107u8
            && module_path[8usize] == 101u8
            && module_path[9usize] == 114u8
            && module_path[10usize] == 58u8
            && module_path[11usize] == 58u8
            && module_path[12usize] == 97u8
            && module_path[13usize] == 99u8
            && module_path[14usize] == 116u8
            && module_path[15usize] == 111u8
            && module_path[16usize] == 114u8
            && module_path[17usize] == 115u8
            && module_path[18usize] == 58u8
            && module_path[19usize] == 58u8
            && module_path[20usize] == 100u8
            && module_path[21usize] == 105u8
            && module_path[22usize] == 115u8
            && module_path[23usize] == 112u8
            && module_path[24usize] == 108u8
            && module_path[25usize] == 97u8
            && module_path[26usize] == 121u8
            && if 27usize == module_path.len() {
                true
            } else {
                module_path[27usize] == b':'
            }
    } {
        return true;
    }
    if if 10usize > module_path.len() {
        false
    } else {
        module_path[0usize] == 100u8
            && module_path[1usize] == 105u8
            && module_path[2usize] == 115u8
            && module_path[3usize] == 99u8
            && module_path[4usize] == 111u8
            && module_path[5usize] == 98u8
            && module_path[6usize] == 105u8
            && module_path[7usize] == 107u8
            && module_path[8usize] == 101u8
            && module_path[9usize] == 114u8
            && if 10usize == module_path.len() {
                true
            } else {
                module_path[10usize] == b':'
            }
    } {
        return true;
    }
    false
}
check()
```
which ends up being 2.5KB. Those 2.5KB log calls add up, fast, and my final debug binary ends up being 30% `check()` functions!

It looks like this code is being generated by `defmt_macros::function_like::log::env_filter::path_check`, which has the comment "Builds a compile-time check that returns `true` when `module_path!` can emit logs at the "requested log `level`". That may be true in optimized builds, but in a regular debug build this all gets compiled and executed at runtime.

@jonas-schievink said:
> This should emit a const item, not a const fn, which would guarantee compile-time evaluation

What do you think about just wrapping the current version in a `const CHECK: bool`? It does compile, but I don't know how to test if it actually solves the problem.

```rust
// current
if {
    const fn check() -> bool {
        if we_want_to_emit_log() {
            return true;
        }
        false
    }
    check()
} {
    emit_defmt_log();
}

// new
if {
    const CHECK: bool = {
        const fn check() -> bool {
            if we_want_to_emit_log() {
                return true;
            }
            false
        }
        check()
    };
    CHECK
} {
    emit_defmt_log();
}
```

I first tried to get rid of the function altogether, but this gives me an "error[E0572]: return statement outside of function body":
```rust
if {
    const CHECK: bool = {
        if we_want_to_emit_log() {
            return true;
        }
        false
    };
    CHECK
} {
    emit_defmt_log();
}
```
---
(All options are in this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d82fe98724afa36aeb244d8d80c399ce)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+- [#xxx]: Wrap const fn in const item to ensure compile-time-evaluation.
 - [#690]: Satisfy clippy
 - [#688]: Release `defmt-decoder 0.3.3`
 - [#687]: `CI`: Re-enable `qemu-snapshot (nightly)` tests
@@ -18,6 +19,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - [#679]: Add changelog enforcer
 - [#678]: Satisfy clippy
 
+[#xxx]: https://github.com/knurling-rs/defmt/pull/xxx
 [#690]: https://github.com/knurling-rs/defmt/pull/690
 [#688]: https://github.com/knurling-rs/defmt/pull/688
 [#687]: https://github.com/knurling-rs/defmt/pull/687

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
-- [#xxx]: Wrap const fn in const item to ensure compile-time-evaluation.
+- [#692]: Wrap const fn in const item to ensure compile-time-evaluation.
 - [#690]: Satisfy clippy
 - [#688]: Release `defmt-decoder 0.3.3`
 - [#687]: `CI`: Re-enable `qemu-snapshot (nightly)` tests
@@ -19,7 +19,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - [#679]: Add changelog enforcer
 - [#678]: Satisfy clippy
 
-[#xxx]: https://github.com/knurling-rs/defmt/pull/xxx
+[#692]: https://github.com/knurling-rs/defmt/pull/692
 [#690]: https://github.com/knurling-rs/defmt/pull/690
 [#688]: https://github.com/knurling-rs/defmt/pull/688
 [#687]: https://github.com/knurling-rs/defmt/pull/687

--- a/macros/src/function_like/log/env_filter.rs
+++ b/macros/src/function_like/log/env_filter.rs
@@ -112,13 +112,15 @@ impl EnvFilter {
             .collect::<Vec<_>>();
 
         Some(quote!({
-            const fn check() -> bool {
-                let module_path = module_path!().as_bytes();
-                #(#checks)*
-                false
-            }
-
-            check()
+            const CHECK: bool = {
+                const fn check() -> bool {
+                    let module_path = module_path!().as_bytes();
+                    #(#checks)*
+                    false
+                }
+                check()
+            };
+            CHECK
         }))
     }
 

